### PR TITLE
Retrieve SSH keys from IMDS first with OVF as a fallback

### PR DESCRIPTION
**Pull request by trstringer**
Currently for Azure we get the SSH keys from the OVF file. This PR adds the feature to attempt to get the SSH keys from IMDS first if available, and only use the OVF source if the data is unavailable from IMDS.

### CODE REVIEW

File: cloudinit/sources/DataSourceAzure.py [hunk 1]
```diff
@@ -559,6 +559,29 @@ def _get_data(self):
...
+        ssh_keys = []
+        try:
+            ssh_keys = [
+                public_key['keyData']
+                for public_key
+                in self.metadata['imds']['compute']['publicKeys']
+            ]
+            LOG.debug('Retrieved SSH keys from IMDS')
+        except KeyError:
+            LOG.debug('Unable to get keys from IMDS, falling back to OVF')
```
Comment (anhvoms) [inline cloudinit/sources/DataSourceAzure.py:574]: msg = 'Unable to get keys from IMDS, falling back to OVF'
LOG.debug(msg)
report_diagnostic_event(msg)

File: cloudinit/sources/DataSourceAzure.py [hunk 2]
```diff
@@ -559,6 +559,29 @@ def _get_data(self):
...
+            report_diagnostic_event('unable to get ssh keys from IMDS')
+            try:
+                ssh_keys = self.metadata['public-keys']
+                LOG.debug('Retrieved keys from OVF')
+            except KeyError:
+                LOG.debug('No keys available from OVF')
```
Comment (anhvoms) [inline cloudinit/sources/DataSourceAzure.py:580]: Same as above

**Comment by jasonzio** [comment]
Other than Anh's comments, which I support, this looks good to me.

**Comment by trstringer** [comment]
Thanks for the review, @smoser!

The desire to get the keys from IMDS (which was not a capability before this PR) is for performance (the keys that we get from IMDS are as-is, whereas what we get from the OVF are fingerprints that we need to make a request to retrieve, decrypt, and then search) as well as simplification (we are moving to a model where IMDS should be the single interface for provisioning).

So hopefully that explains why we want to get the keys from IMDS ideally. But we have to support the fallback to OVF is that not all environments have IMDS yet (like Azure Stack).

Does that answer your questions?

File: cloudinit/sources/DataSourceAzure.py [hunk 3]
```diff
@@ -1431,7 +1471,7 @@ def get_metadata_from_imds(fallback_nic, retries):
 @azure_ds_telemetry_reporter
 def _get_metadata_from_imds(retries):
 
-    url = IMDS_URL + "instance?api-version=2017-12-01"
+    url = IMDS_URL + "instance?api-version=2019-06-01"
```
Comment (smoser) [inline cloudinit/sources/DataSourceAzure.py:1474]: is azure-stack guaranteed to have this api version?

Comment (trstringer) [reply cloudinit/sources/DataSourceAzure.py:1474]: Whether Azure Stack doesn't have that version, or no IMDS at all, a failure is a failure here. The situation is the same. The only consideration is are we causing an issue for existing Azure Stack implementations that *do* have IMDS but do *not* have api-version 2019-06-01. I believe that this is not the case or a concern. @anhvoms, let me know if you disagree with that statement.

Comment (anhvoms) [reply cloudinit/sources/DataSourceAzure.py:1474]: There is no IMDS support in Azure Stack currently. Once they do I would expect them to pick up the latest version

Comment (trstringer) [reply cloudinit/sources/DataSourceAzure.py:1474]: Great thanks for the clarification, @anhvoms. So it sounds like this won't be a breaking change for Azure Stack then.

**Comment by jasonzio** [comment]
@smoser - a key scenario is support VMs with cloud-init that don't have a dvd device installed in the kernel. We regularly deal with developers who've received security guidance telling them to remove the dvd device; we'd like that to just work, rather than having to argue with them about the quality of the guidance.

**Comment by smoser** [comment]
I think that I'd like it documented somewhere the reason for the "fallback to OVF" path, and that in the new IMDS case on azure, there may not even be a DVD.

That isn't obvious at all, and anyone other than the 2 or 3 of you here that are looking at it wont' know.

I don't have issues with any of the general codepath in the PR just in being able to understand the environment that this is expected to run in.

**Comment by trstringer** [comment]
Makes sense!! I'll get those changes pushed soon. Thanks!

**Comment by anhvoms** [comment]
azure.py: L562
You should switch the order of this check:
"if goal_state.certificates_xml is not None and pubkey_info is not None:"

This will skip the call to retrieve the certificate data that has the ssh public keys (this property triggers a call to wireserver populate certificates_xml when it's not yet initialized)

Also, you might be able to optimize this further by skipping the generation of the transport cert (which is generated by cloud-init and used by the platform to encrypt the certificate data)

**Comment by trstringer** [comment]
Documentation added and a few other improvements to the code as well! Let me know your thoughts, @smoser, and if you think anything else is needed. // Cc @anhvoms

**Comment by TheRealFalcon** [comment]
Hey, @trstringer , I noticed you recently marked this PR as "draft". Does that mean you're still making changes and are no longer currently looking for review?

**Comment by trstringer** [comment]
@TheRealFalcon, that's correct! Just doing some final testing on this PR before it's ready to get merged. Thanks!

**Comment by trstringer** [comment]
@mitechie Testing complete!

**Comment by mitechie** [comment]
Awesome thank you. We'll get it in the queue.

**Review by lucasmoura — APPROVED** [review]

**Review by mitechie — APPROVED** [review]

**Pull request merged**